### ISSUE TITLE
Crash when exiting directly while a filter UI menu is active

### ISSUE THREAD

**Issue by katananja** [mamedev/mame#4596]
``MAME v0.206 (mame0206-104-g55ead0c997)``
Debian 9.7 x64
Compile options:
``make clean && make SYMBOLS=1 SYMLEVEL=1 OPTIMIZE=0``

1. Run mame ``./mame64 -window -maximize``
2. Select Category, it will show the default message "_No category INI files found_"
3. **Close the window**
4. Segmentation fault happen.

```
Thread 1 "mame64" received signal SIGSEGV, Segmentation fault.
─── Assembly ───────────────────────────────────────────────────────────────────
0x00007ffff4e09d70 ? mov    %rdi,%rax
0x00007ffff4e09d73 ? mov    (%rdi),%rdi
0x00007ffff4e09d76 ? add    $0x10,%rax
0x00007ffff4e09d7a ? cmp    %rax,%rdi
0x00007ffff4e09d7d ? je     0x7ffff4e09d88 <_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEED2Ev+24>
─── Expressions ────────────────────────────────────────────────────────────────
─── History ────────────────────────────────────────────────────────────────────
─── Memory ─────────────────────────────────────────────────────────────────────
─── Registers ──────────────────────────────────────────────────────────────────
   rax 0x00007fffd0d38030     rbx 0x0000000000000007     rcx 0x00007fffd1036bd0 
   rdx 0x00007fffd1036bd0     rsi 0x00007fffd1036bd0     rdi 0x00007fffd0d38030 
   rbp 0x00007fffffff6d50     rsp 0x00007fffffff6d38      r8 0x000000000000001c 
    r9 0x00007ffff455fed0     r10 0x6f7669757172615b     r11 0x657320494e492073 
   r12 0x0000555557d6d790     r13 0x00007fffffffde20     r14 0x0000000000000000 
   r15 0x0000000000000000     rip 0x00007ffff4e09d73  eflags [ PF IF RF ]       
    cs 0x00000033              ss 0x0000002b              ds 0x00000000         
    es 0x00000000              fs 0x00000000              gs 0x00000000         
─── Source ─────────────────────────────────────────────────────────────────────
─── Stack ──────────────────────────────────────────────────────────────────────
[0] from 0x00007ffff4e09d73 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()
(no arguments)
[1] from 0x000055555dc249ca in ui::menu_item::~menu_item at ../../../../../src/osd/../frontend/mame/ui/menuitem.h:30
(no arguments)
[+]
─── Threads ────────────────────────────────────────────────────────────────────
[8] id 29872 name SDLAudioDev1 from 0x00007ffff4511741 in __GI_ppoll+145 at ../sysdeps/unix/sysv/linux/ppoll.c:39
[7] id 29871 name PulseHotplug from 0x00007ffff4511741 in __GI_ppoll+145 at ../sysdeps/unix/sysv/linux/ppoll.c:39
[6] id 29870 name si_shader:3 from 0x00007ffff74a115f in pthread_cond_wait@@GLIBC_2.3.2 at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
[5] id 29869 name si_shader:2 from 0x00007ffff74a115f in pthread_cond_wait@@GLIBC_2.3.2 at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
[4] id 29868 name si_shader:1 from 0x00007ffff74a115f in pthread_cond_wait@@GLIBC_2.3.2 at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
[3] id 29867 name si_shader:0 from 0x00007ffff74a115f in pthread_cond_wait@@GLIBC_2.3.2 at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
[2] id 29866 name radeon_cs:0 from 0x00007ffff74a115f in pthread_cond_wait@@GLIBC_2.3.2 at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
[1] id 29733 name mame64 from 0x00007ffff4e09d73 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()
────────────────────────────────────────────────────────────────────────────────
0x00007ffff4e09d73 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
>>> where
#0  0x00007ffff4e09d73 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x000055555dc249ca in ui::menu_item::~menu_item () at ../../../../../src/osd/../frontend/mame/ui/menuitem.h:30
#2  0x000055555dc249f1 in std::_Destroy<ui::menu_item> () at /usr/include/c++/6/bits/stl_construct.h:93
#3  0x000055555dc243d5 in std::_Destroy_aux<false>::__destroy<ui::menu_item*> () at /usr/include/c++/6/bits/stl_construct.h:103
#4  0x000055555dc23bcc in std::_Destroy<ui::menu_item*> () at /usr/include/c++/6/bits/stl_construct.h:126
#5  0x000055555dc23391 in std::_Destroy<ui::menu_item*, ui::menu_item> () at /usr/include/c++/6/bits/stl_construct.h:151
#6  0x000055555dc23484 in std::vector<ui::menu_item, std::allocator<ui::menu_item> >::_M_erase_at_end () at /usr/include/c++/6/bits/stl_vector.h:1436
#7  0x000055555dc22b22 in std::vector<ui::menu_item, std::allocator<ui::menu_item> >::clear () at /usr/include/c++/6/bits/stl_vector.h:1210
#8  0x000055555e0e7940 in ui::menu::reset () at ../../../../../src/frontend/mame/ui/menu.cpp:288
#9  0x000055555dd32c82 in operator() () at ../../../../../src/frontend/mame/ui/selgame.cpp:1421
#10 0x000055555dd33b55 in _M_invoke () at /usr/include/c++/6/functional:1731
#11 0x000055555ddb7ba9 in std::function<void (ui::machine_filter&)>::operator()(ui::machine_filter&) const () at /usr/include/c++/6/functional:2127
#12 0x000055555dd9316d in ~menu_configure () at ../../../../../src/frontend/mame/ui/utils.cpp:960
#13 0x000055555dd931be in ~menu_configure () at ../../../../../src/frontend/mame/ui/utils.cpp:961
#14 0x000055555dd403ac in std::default_delete<ui::menu>::operator() () at /usr/include/c++/6/bits/unique_ptr.h:76
#15 0x000055555e0edc23 in std::unique_ptr<ui::menu, std::default_delete<ui::menu> >::reset () at /usr/include/c++/6/bits/unique_ptr.h:347
#16 0x000055555e0ed01e in std::unique_ptr<ui::menu, std::default_delete<ui::menu> >::operator=(std::unique_ptr<ui::menu, std::default_delete<ui::menu> >&&) () at /usr/include/c++/6/bits/unique_ptr.h:254
#17 0x000055555e0e7370 in ui::menu::global_state::clear_free_list () at ../../../../../src/frontend/mame/ui/menu.cpp:166
#18 0x000055555e0e7570 in ui::menu::exit () at ../../../../../src/frontend/mame/ui/menu.cpp:211
#19 0x0000555558d24d2d in delegate_base<void>::operator() () at ../../../../../src/lib/util/delegate.h:531
#20 0x0000555561a01205 in running_machine::call_notifiers () at ../../../../../src/emu/machine.cpp:884
#21 0x00005555619ff963 in running_machine::run () at ../../../../../src/emu/machine.cpp:415
#22 0x000055555dd0f88e in mame_machine_manager::execute () at ../../../../../src/frontend/mame/mame.cpp:239
#23 0x000055555dddf423 in cli_frontend::start_execution () at ../../../../../src/frontend/mame/clifront.cpp:262
#24 0x000055555dddf68f in cli_frontend::execute () at ../../../../../src/frontend/mame/clifront.cpp:278
#25 0x000055555dd1023f in emulator_info::start_frontend () at ../../../../../src/frontend/mame/mame.cpp:339
#26 0x000055555dc1757b in main () at ../../../../../src/osd/sdl/sdlmain.cpp:217
```

**Comment by cuavas** [comment]
Fixed in 0470ded12022841d9c59941598cfbc31427d3d14 - path of least resistance is just to reverse the free list again before unwinding it.  No extra structures, no additional points needing O(n) operations.

**Comment by katananja** [comment]
Thank you @cuavas.
